### ISSUE TITLE
moskito lifecycle? UpdateTriggerServiceImpl opens a thread "MoskitoIntervalUpdater" which remains after tomcat undeploy

### ISSUE THREAD

**Issue by FelixJongleur42** [anotheria/moskito#218]
This for **hot deployment** under tomcat (and possibly other application servers) with multiple deployments of the same war where moskito is in the classpath of the war, not the server, because we want individual stats for each war.

**There should be some mechanism that "tears down" moskito in a servlet context when the servlet context is destroyed?**

I see this candidate for this: net.anotheria.moskito.webui.util.StartStopListener

Somehow everything is started via class loading and <clinit> stuff, but there should be some lifecycle like in Spring with DisposableBean interface?

e.g. **IUpdateTriggerService should have a method cleanup() which cancels the timer**.

net.anotheria.moskito.webui.util.StartStopListener already does something similar for:

```
	@Override
	public void contextDestroyed(ServletContextEvent sce) {
		log.info("Deinitializing moskito for "+sce.getServletContext().getServletContextName());
		BuiltinUpdater.cleanup();
		ThresholdRepository.getInstance().cleanup();
	}

```

The warning tomcat gives you is like:

> 20-Dec-2018 17:02:38.671 WARNUNG [ContainerBackgroundProcessor[StandardEngine[Catalina]]] org.apache.catalina.loader.WebappClassLoaderBase.clearReferencesThreads The web application [XXX] appears to have started a thread named [MoskitoIntervalUpdater] but has failed to stop it. This is very likely to create a memory leak. Stack trace of thread:
>  java.lang.Object.wait(Native Method)
>  java.util.TimerThread.mainLoop(Timer.java:552)
>  java.util.TimerThread.run(Timer.java:505)

I double-checked: this message goes away when declaring StartStopListener which calls BuiltinUpdater.cleanup();

> 20-Dec-2018 17:02:38.675 WARNUNG [ContainerBackgroundProcessor[StandardEngine[Catalina]]] org.apache.catalina.loader.WebappClassLoaderBase.clearReferencesThreads The web application [XXX] appears to have started a thread named [MoSKito Builtin-Updater] but has failed to stop it. This is very likely to create a memory leak. Stack trace of thread:
>  java.lang.Object.wait(Native Method)
>  java.util.TimerThread.mainLoop(Timer.java:552)
>  java.util.TimerThread.run(Timer.java:505)

**Comment by dvayanu** [comment]
Hello FelixJongleur42,

so if you declare the start/stop listener, the only remaining issue is the intervalupdater-timer?

regards
Leon

**Comment by FelixJongleur42** [comment]
> the only remaining issue is the intervalupdater-timer?

Yes, for core that is the only problematic thread I could identify.
There's another Thread though, but it's from ConfigureMe.

```
11-Dec-2017 12:12:03.769 WARNING [localhost-startStop-2] org.apache.catalina.loader.WebappClassLoaderBase.clearReferencesThreads The web application [soo] appears to have started a thread named [Thread-8] but has failed to stop it. This is very likely to create a memory leak. Stack trace of thread:
 java.lang.Thread.sleep(Native Method)
 org.configureme.sources.ConfigurationSourceRegistry$WatcherThread.run(ConfigurationSourceRegistry.java:164)
```

Source
https://github.com/anotheria/configureme/blob/master/src/main/java/org/configureme/sources/ConfigurationSourceRegistry.java

I filed a separate issue on that:
https://github.com/anotheria/configureme/issues/12

**Comment by dvayanu** [comment]
is it possible for you to test with a snapshot version? I've tested with 2.8.10-SNAPSHOT (will be released as 2.9.0 due to changes in the dashboard api semantics) which includes fixes for configureme#12 and I don't see neither of the messages.
I see though following message (tomcat8):
01-Jan-2019 02:02:29.538 WARNING [localhost-startStop-1] org.apache.catalina.loader.WebappClassLoaderBase.clearReferencesThreads The web application [burgershop] appears to have started a thread named [pool-1-thread-1] but has failed to stop it. This is very likely to create a memory leak. Stack trace of thread:
 sun.misc.Unsafe.park(Native Method)
 java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
 java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
 java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1093)
 java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:809)
 java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
 java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
 java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
 java.lang.Thread.run(Thread.java:748)


From what I see, I am not quite sure, it's a thread in moskito at all. according to jstack it seems not:
"MoSKito Builtin-Updater" #21 daemon prio=5 os_prio=31 tid=0x00007feb901f9000 nid=0xa003 in Object.wait() [0x000070000d417000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000005c00e21f0> (a java.util.TaskQueue)
	at java.util.TimerThread.mainLoop(Timer.java:552)
	- locked <0x00000005c00e21f0> (a java.util.TaskQueue)
	at java.util.TimerThread.run(Timer.java:505)

"pool-1-thread-1" #20 prio=5 os_prio=31 tid=0x00007feb8f33e000 nid=0x6303 waiting on condition [0x000070000d314000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000005c00d2808> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1081)
	at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:809)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

"ConfigurationSourceRegistry.WatcherThread" #19 daemon prio=5 os_prio=31 tid=0x00007feb97be9000 nid=0xa203 waiting on condition [0x000070000d211000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.configureme.sources.ConfigurationSourceRegistry$WatcherThread.run(ConfigurationSourceRegistry.java:179)

"MoskitoIntervalUpdater" #18 daemon prio=5 os_prio=31 tid=0x00007feb927f1800 nid=0xa303 in Object.wait() [0x000070000d10e000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000005c00e2540> (a java.util.TaskQueue)
	at java.util.TimerThread.mainLoop(Timer.java:552)
	- locked <0x00000005c00e2540> (a java.util.TaskQueue)
	at java.util.TimerThread.run(Timer.java:505)

**Comment by dvayanu** [comment]
i added shutdown methods on AlertDispatcher and the log output on tomcat 8 is now pretty uneventful, can you test prior to a release?

**Comment by FelixJongleur42** [comment]
> i added shutdown methods on AlertDispatcher

as far as I can recall we where not even aware of any AlertDispatcher - maybe this is because we are using an older version of moskito (not in the office right now - so I cannot check) or the alert dispatcher was not activated in some way or it was implemented differently in an older version?

I will have to check when I get back to the office (next week).

**Comment by dvayanu** [comment]
fixed in 2.9.0
